include/seqan3/core/detail/pack_algorithm.hpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides algorithms for meta programming, parameter packs and seqan3::type_list.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <seqan3/core/type_list/type_list.hpp>
+#include <seqan3/core/tuple_utility.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/type_traits>
+
+namespace seqan3::detail
+{
+
+//!\cond
+template <typename list_t>
+struct type_identity_delegate;
+//!\endcond
+
+/*!\brief Helper invocation delegate fot type based meta algorithms.
+ * \tparam list_t The template template argument whose contained types should be invoked with the given function.
+ * \tparam pack_t The template parameter pack to apply the target function on.
+ *
+ * \details
+ *
+ * The meta algorithms provide a pack version and a type based version. In the type based version the algorithm is
+ * called on the types contained in the enclosing list type after they have been wrapped in std::type_identity.
+ * This requires to expand the types of the list type in a helper function since it cannot be done inside of the
+ * function template declaration. To avoid adding an additional impl function for every algorithm, which takes care of
+ * the unpacking of the parameters, this helper struct provides an invocation method to pass the wrapped parameters
+ * to the desired parameter pack version of the algorithm. Since the target function is iteself a function template
+ * we cannot simply pass a pointer to that function without knowing the types of it's argument before. Accordingly,
+ * one needs to defer the invocation in a lambda function which is called in the static
+ * seqan3::detail::type_identity_delegate::invoke method.
+ */
+template <template <typename ...> typename list_t, typename ...args_t>
+struct type_identity_delegate<list_t<args_t...>>
+{
+    /*!\brief Invokes the actual function by passing the types as instances of std::type_identity to the target
+     *        function.
+     * \tparam wrapper_t The type of the function wrapper.
+     * \param[in] wrapper The wrapper function, e.g. a lambda function.
+     * \returns The std::invoke_result of calling wrapper with the parameters.
+     *
+     * \details
+     *
+     * Invokes the wrapper by passing the types wrapped in std::type_identity.
+     */
+    template <typename wrapper_t>
+    //!\cond
+        requires std::invocable<wrapper_t, std::type_identity<args_t>...>
+    //!\endcond
+    static constexpr std::invoke_result_t<wrapper_t, std::type_identity<args_t>...> invoke(wrapper_t && wrapper)
```

Comment:
well, I called it wrapper because it wraps the function that should be actually invoked. Since it is a template function I cannot pass the function by reference but had to use a lambda that then calls the actual function. Thus I figured fn_t is a little unspecific.

Reply:
I did rename it, because I realised that also with the naming of wrapper_t I was actually pursuing the specific case of the functions we had so far, but that does not necessarily need to be. So I adapted the naming and the documentation accordingly.